### ISSUE TITLE
Use Balsa HTTP/1 codec

### ISSUE THREAD

**Issue by bencebeky** [envoyproxy/envoy#21245]
The current HTTP/1 codec is [http-parser](https://github.com/nodejs/http-parser).  The plan is to transition Envoy to use [Balsa](https://github.com/google/quiche/tree/main/quiche/balsa) instead.  [edit: fix link]

**Comment by daixiang0** [comment]
@bencebeky hi, could you compare Balsa with http-parser to show the advantage of transition?

**Comment by wbpcode** [comment]
Yeah, is there any reason to make this transition?

**Comment by adisuissa** [comment]
> 

This is related to #10988.
The current http_parser is no longer maintained and adding features such as #18819 is non-trivial.

**Comment by alyssawilk** [comment]
Ah, I should have cross-linked when I was triaging.  The decision points largely documented at https://docs.google.com/document/d/1Z8neuR6GTL61EU8jC88SD2as6Nmq15B-BNWC-SBVKns/edit after a couple of rounds of discussion at the community meeting (notes here https://docs.google.com/document/d/1Yc4zkV-A_cC_R3C0u6D15WQAsRSxADs8p2l8UMMa4P4/edit# as it's not convenient for all time zones)

**Comment by kanongil** [comment]
FYI, the balsa link is broken. They *just* moved it to: https://github.com/google/quiche/tree/main/quiche/balsa.

**Comment by bencebeky** [comment]
no stalebot

**Comment by bencebeky** [comment]
Note that BalsaParser is being developed to provide feature (and bug) parity with http-parser to the highest reasonable extent.  This includes identical error messages, some of which have "HPE_" prefix (for "http-parser error"), and some of which are hardcoded in an awkward manner, see #22447 for an example.  This is to minimize friction during the transition.  After the transision is complete, these error messages, along with other peculiarities, can be eliminated by behavior-changing PR, using runtime flag protection as appropriate.

**Comment by JuniorHsu** [comment]
> Ah, I should have cross-linked when I was triaging. The decision points largely documented at https://docs.google.com/document/d/1Z8neuR6GTL61EU8jC88SD2as6Nmq15B-BNWC-SBVKns/edit 

@alyssawilk This doc is not public or doesn't exist

**Comment by alyssawilk** [comment]
or @yanavlasov codec analysis?  I can't recall :-/

**Comment by JuniorHsu** [comment]
We're keen to do some verification on balsa however black box testing is risky for us. Some document is helpful for us to see the impact/possible regression

**Comment by alyssawilk** [comment]
@bencebeky can you share the internal docs of known differences (i.e. lack thereof?)  I think it largely came down to the one we just fixed, and slight difference in max header byte accounting.

the tl;dr for migration was http_parser is abandonware, getting fixes to security issues has proven quite difficult, getting enhancements largely implausible, so the Envoy maintainers agreed migration was desirable.
For why balsa, largely because it's been deployed at scale in Google for 15 years, and because Google was willing to do the work to migrate off a deprecated parser if it was one we both trusted, and knew wouldn't be abandoned, and no one else offered up cycles to do the port :-)

We've been running balsa internally for a month in some internal deployments for about a month now with no issue.  

Happy to answer any other questions you might have on github/slack/vc :-)

**Comment by bencebeky** [comment]
Hi. Here's the public fork of the document discussing behavioral differences: https://docs.google.com/document/d/1FSpeF7QlmkQ4JhfpKWzEL0UKM3OLTTnd1VJJtnFlq64.

We have also done load testing manually, and found no significant difference in memory or CPU usage. Also did some testing with synthetic data to find differences in behavior, but found none. Fuzz testing found the extended ASCII issue, which is being fixed.

**Comment by JuniorHsu** [comment]
Thanks. Those help a lot and we'll move on some internal verification :) Would post if anything is concerned

**Comment by alyssawilk** [comment]
awesome.  please update either way, as once you folks have tested as well we'll flip it true by default :-)

**Comment by andreycpp** [comment]
Hi @bencebeky. I've recently bumped into the lack of custom methods support issue #18819 in one of our prod environments. Interestingly http-parser
[does support some custom methods](https://github.com/nodejs/http-parser/blob/ec8b5ee63f0e51191ea43bb0c6eac7bfbff3141d/http_parser.h#L165C1-L208C1) but they are all hardcoded. 

It looks like the solution to custom methods is to migrate to this Balsa library. Could you please clarify where is this work, and when is it expected to land? Is there possibly a way to use Balsa even today (maybe with some limitations/corner cases)? Thanks!

**Comment by wbpcode** [comment]
Iirc, balsa is enabled by default after envoy 1.28.

**Comment by bencebeky** [comment]
Hi, as noted above, Balsa is default enabled. However, this is only the first step towards allowing arbitrary methods. BalsaParser has the list of allowed methods copied over verbatim from http-parser, see https://github.com/envoyproxy/envoy/blob/2970ddbd4ade787dd51dfbe605ae2e8c5d8ffcf7/source/common/http/http1/balsa_parser.cc#L54-L59. The two parsers behave identically to minimize friction due to migrating from one parser to the other.

There's a related project called UHV ("universal header validator" or "unified header validation") that moves validation logic, including method validation, from the individual HTTP/1, HTTP/2, HTTP/3 codec implementations to a centralized place. It will provide a configuration setting for enabling custom methods.

**Comment by andreycpp** [comment]
Thanks @bencebeky!

Would envoy community accept addition of new methods to Balsa? The method we need right now is the BITS_POST which is used by [BITS Upload Protocol](https://learn.microsoft.com/en-us/windows/win32/bits/bits-upload-protocol]). It acts exactly like POST, just has different name.

The UHV work - is it tracked via https://github.com/envoyproxy/envoy/issues/10646?

**Comment by alyssawilk** [comment]
Yep, that's the correct issue for UHV.  You should be able to test with UHV on using envoy.reloadable_features.enable_universal_header_validator though there may be some minor differences not yet runtime guarded.  I believe once you're using balsa and uhv you can follow thorugh on the TODO to complete defer method validation ot UHV and extend UHV to allow custom methods.  Happy to do the review along with @yanavlasov

**Comment by birenroy** [comment]
The migration from http-parser to Balsa seems to have stuck, so I think this issue can be closed. (If there are other HTTP/1-related feature requests, perhaps they could be opened as separate issues?)

@RyanTheOptimist

**Comment by ggreenway** [comment]
I thought balsa was the default http1 parser at this time. Isn't this now complete?

**Comment by RyanTheOptimist** [comment]
> I thought balsa was the default http1 parser at this time. Isn't this now complete?

Yes, I think that's what @birenroy was trying to say. Balsa is the default HTTP/1 parser and has been for a while without problem. So we can close this issue as complete.

**Comment by birenroy** [comment]
> Yes, I think that's what [@birenroy](https://github.com/birenroy) was trying to say. Balsa is the default HTTP/1 parser and has been for a while without problem. So we can close this issue as complete.

Yes, sorry for the confusing wording. I meant that Balsa has been made the default, and has been working as intended for a while now without being reverted.

**Comment by ggreenway** [comment]
Sorry, I read that wrong; in my head I read `The migration from http-parser to Balsa seems stuck`. 

I think we can close this as complete, as @RyanTheOptimist said.

**Comment by ravenblackx** [comment]
There is a TODO referencing this issue that hasn't been done (or at least the TODO hasn't been removed), [here](https://github.com/envoyproxy/envoy/blob/3c9f5fca3f1769b0dd5c60bd82dbce4135cfc5b0/source/common/http/http1/balsa_parser.cc#L36-L37). Given how dreadful the related code is, it would be nice for the TODO to be done.

**Comment by bencebeky** [comment]
> There is a TODO referencing this issue that hasn't been done (or at least the TODO hasn't been removed), [here](https://github.com/envoyproxy/envoy/blob/3c9f5fca3f1769b0dd5c60bd82dbce4135cfc5b0/source/common/http/http1/balsa_parser.cc#L36-L37). Given how dreadful the related code is, it would be nice for the TODO to be done.

I agree. @yanavlasov What is the status of UHV?
